src/gen/run_block_generator.rs

```diff
@@ -75,3 +78,111 @@ pub fn run_block_generator<GenBuf: AsRef<[u8]>>(
     result.cost += max_cost - cost_left;
     Ok(result)
 }
+
+fn extract_n<const N: usize>(
+    a: &Allocator,
+    mut n: NodePtr,
+    e: ErrorCode,
+) -> Result<[NodePtr; N], ValidationErr> {
+    let mut ret: [NodePtr; N] = [0; N];
+    let mut counter = 0;
+    assert!(N > 0);
+    while let Some((item, rest)) = a.next(n) {
+        if counter == N - 1 {
+            break;
+        }
+        n = rest;
+        ret[counter] = item;
+        counter += 1;
+    }
+    if counter != N - 1 {
+        return Err(ValidationErr(n, e));
+    }
+    ret[counter] = n;
+    Ok(ret)
+}
+
+// This has the same behavior as run_block_generator() but implements the
+// generator ROM in rust instead of using the CLVM implementation.
+// it is not backwards compatible in the CLVM cost computation (in this version
+// you only pay cost for the generator, the puzzles and the conditions).
+// it also does not apply the stack depth or object allocation limits the same,
+// as each puzzle run in its own environment.
+pub fn run_block_generator2<GenBuf: AsRef<[u8]>>(
```

Comment:
Maybe `run_block_generator_native` or `run_block_generator_rust`?

Reply:
in my mind, the main property of this function is that it doesn't charge cost for the ROM. Everything else is an implementation detail. I thought `2` was a reasonable name because it's version 2 and the hard fork is even introduced in `2.0`